src/analysis/stack-lattice.h

```diff
@@ -0,0 +1,139 @@
+#ifndef wasm_analysis_stack_lattice_h
+#define wasm_analysis_stack_lattice_h
+
+#include <deque>
+
+#include "lattice.h"
+
+namespace wasm::analysis {
+
+template<typename StackFrameLattice> class StackLattice {
+  static_assert(is_lattice<StackFrameLattice>);
+  StackFrameLattice& stackFrameLattice;
+
+public:
+  StackLattice(StackFrameLattice& stackFrameLattice)
+    : stackFrameLattice(stackFrameLattice) {}
+
+  class Element {
+    std::deque<typename StackFrameLattice::Element> stackValue;
+    bool top = false;
+
+  public:
+    Element() = default;
+
+    Element(Element&& source) = default;
+    Element(Element& source) = default;
+
+    Element& operator=(Element&& source) = default;
+    Element& operator=(const Element& source) = default;
+
+    bool isTop() { return top; }
+    void setTop(bool val) { top = val; }
+    bool isBottom() { return !isTop && stackValue.empty(); }
+
+    typename StackFrameLattice::Element& stackTop() {
+      return stackValue.back();
+    }
+    void push(typename StackFrameLattice::Element frame) {
+      stackValue.push_back(frame);
+    }
+    void pop() { stackValue.pop_back(); }
+
+    bool makeLeastUpperBound(const Element& other) {
+      bool modified = false;
+      auto otherIterator = other.stackValue.crbegin();
+      for (auto thisIterator = stackValue.rbegin();
+           thisIterator != stackValue.rend();
+           ++thisIterator) {
```

Comment:
You could check `otherIterator != other.stackValue.crend()` here in the loop condition and I believe you could do `++thisIterator, ++otherIterator` as the update statement to make this really nice and symmetrical. If that does work, I would pull out `thisIterator` and initialize it above the loop for symmetry with `otherIterator` as well.